stan/math/opencl/rev/log_diff_exp.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef STAN_MATH_OPENCL_REV_LOG_DIFF_EXP_HPP
+#define STAN_MATH_OPENCL_REV_LOG_DIFF_EXP_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta/is_kernel_expression.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/opencl/prim/sum.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/rev/core/reverse_pass_callback.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the natural logarithm of the difference of the
+ * inverse logits of the specified arguments.
+ *
+ * @tparam T_x type of x argument
+ * @tparam T_y type of y argument
+ * @param x first argument
+ * @param y second argument
+ * @return Result of log difference of inverse logits of arguments.
+ */
+template <
+    typename T_x, typename T_y,
+    require_all_nonscalar_prim_or_rev_kernel_expression_t<T_x, T_y>* = nullptr,
+    require_any_var_t<T_x, T_y>* = nullptr>
+inline var_value<matrix_cl<double>> log_diff_exp(T_x&& x, T_y&& y) {
+  arena_t<T_x> x_arena = std::forward<T_x>(x);
+  arena_t<T_y> y_arena = std::forward<T_y>(y);
+
+  matrix_cl<double> res_val
+      = log_diff_exp(value_of(x_arena), value_of(y_arena));
+
+  return make_callback_var(
+      res_val,
+      [x_arena, y_arena](const vari_value<matrix_cl<double>>& res) mutable {
+        if (!is_constant<T_x>::value && !is_constant<T_y>::value) {
```

Comment:
What is the reason for separete branches here?

Reply:
Because that is how the CPU version is done: https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/log_diff_exp.hpp

I dont know enough about the math behind this and in those cases I assume the CPU version is done correctly and I just want to mimic it.

Reply:
To me it seems that the vv branch should be able to handle all cases. @SteveBronder @bbbales2 Do you have any idea why there is special inf handling in vd and dv cases?

Reply:
I do agree it looks weird. My guess would be that we should probably handle those in vv as well.

Reply:
> there is special inf handling in vd and dv cases

I don't know why it is there and I don't like it.

I was copying some old code recently and Steve pointed out where it was doing something like:
```
bvi_->adj_ -= adj_ * INFTY;
```

Why do we multiply the infinity by infinity? As long as `adj_` is finite that is just infinity. Best you can do is check the git blame on the CPU thing and see if there was a recent issue.

I guess options are:
1. Remove the checks in CPU and GPU to make them consistent
2. Leave the checks out in GPU and assume it won't cause major differences
3. Copy the checks on the GPU code and just roll with it
4. Copy the checks to the GPU code and plan to remove them in the future

Especially since the CPU is inconsistent with itself, I'm happy with anything here.

Reply:
The reason you would multiply with infinity is for edge cases like 0 * Inf = NaN. I dont think that matters in Stan, but might otherwise.

The way its done currently is consistent with the CPU version where vv has no checks and vd and dv has checks.

Reply:
I think the best plan is to keep GPU version consistent with the CPU and make a separate issue to get more eyes on that? @t4c1 what do you think?

Reply:
I would prefer to remove these checks from both versions, but I am fine with either option.